python/triton_mlir/triton_mlir/MaxsMachineFunction.cpp

```diff
@@ -0,0 +1,210 @@
+  }
+
+  void getAnalysisUsage(AnalysisUsage &au) const override;
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+  MachineFunctionProperties getRequiredProperties() const override {
+    return MachineFunctionProperties().set(
+        MachineFunctionProperties::Property::IsSSA);
+  }
+};
+} // end anonymous namespace
+
+namespace llvm {
+extern char &MaxsMachineFunctionID;
+}
+
+char MaxsMachineFunction::ID = 0;
+char &llvm::MaxsMachineFunctionID = MaxsMachineFunction::ID;
+
+static void initializeMaxsMachineFunctionPassOnce(PassRegistry &Registry) {
+  PassInfo *PI = new PassInfo(
+      "MaxsMachineFunction", "maxsmachinefunction", &MaxsMachineFunction::ID,
+      PassInfo::NormalCtor_t(callDefaultCtor<MaxsMachineFunction>), false,
+      false);
+  Registry.registerPass(*PI, true);
+}
+
+static llvm::once_flag InitializeMaxsMachineFunctionPassFlag;
+
+void initializeMaxsMachineFunctionPass(PassRegistry &Registry) {
+  llvm::call_once(InitializeMaxsMachineFunctionPassFlag,
+                  initializeMaxsMachineFunctionPassOnce, std::ref(Registry));
+}
+
+FunctionPass *createMaxsMachineFunctionPass() {
+  return new MaxsMachineFunction();
+}
+
+void MaxsMachineFunction::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.setPreservesCFG();
+  AU.addPreserved<AAResultsWrapperPass>();
+  MachineFunctionPass::getAnalysisUsage(AU);
+}
+
+std::optional<MachineInstr *> findNthUser(MachineInstr &MI,
+                                          MachineRegisterInfo *MRI,
+                                          const Register &CRReg,
+                                          unsigned N = 1) {
+  MachineBasicBlock::iterator I = MI;
+  unsigned Idx = 0;
+  for (MachineBasicBlock::iterator EL = MI.getParent()->end(); I != EL; ++I) {
+    for (MachineRegisterInfo::use_instr_iterator
+             J = MRI->use_instr_begin(CRReg),
+             JE = MRI->use_instr_end();
+         J != JE; ++J)
+      if (&*J == &*I) {
+        Idx++;
+      }
+    if (Idx == N) {
+      return &*I;
+    }
+  }
+  return {};
+}
+
+namespace llvm {
+namespace AMDGPU {
+extern const TargetRegisterClass VReg_64_Align2RegClass;
+extern const TargetRegisterClass VGPR_32RegClass;
+enum {
+  REG_SEQUENCE = 19,
+  V_ADD_F32_e32 = 5304,
+  V_PK_ADD_F32 = 9545,
+};
+
+enum : uint16_t {
+  NoSubRegister,
+  hi16,                                    // 1
+  lo16,                                    // 2
+  sub0,                                    // 3
+  sub0_sub1,                               // 4
+  sub0_sub1_sub2,                          // 5
+  sub0_sub1_sub2_sub3,                     // 6
+  sub0_sub1_sub2_sub3_sub4,                // 7
+  sub0_sub1_sub2_sub3_sub4_sub5,           // 8
+  sub0_sub1_sub2_sub3_sub4_sub5_sub6_sub7, // 9
+  sub0_sub1_sub2_sub3_sub4_sub5_sub6_sub7_sub8_sub9_sub10_sub11_sub12_sub13_sub14_sub15, // 10
+  sub1, // 11
+};
+} // namespace AMDGPU
+} // namespace llvm
```

Comment:
enum copy-pasta from td generated files which currently aren't distributed (but should be distributed just like how MLIR td files are distributed so downstream can rebuild)